File: educational-core/src/com/jetbrains/edu/learning/checkio/options/CheckiOOptions.java
package com.jetbrains.edu.learning.checkio.options;

import com.intellij.ui.HyperlinkAdapter;
import com.jetbrains.edu.learning.checkio.account.CheckiOAccount;
import com.jetbrains.edu.learning.checkio.connectors.CheckiOOAuthConnector;
import com.jetbrains.edu.learning.settings.OauthOptions;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.event.HyperlinkEvent;

public abstract class CheckiOOptions extends OauthOptions<CheckiOAccount> {
  private final CheckiOOAuthConnector myOAuthConnector;

  protected CheckiOOptions(@NotNull CheckiOOAuthConnector oauthConnector) {
    super();
    myOAuthConnector = oauthConnector;
    initAccounts();
  }

  @Nullable
  @Override
  public CheckiOAccount getCurrentAccount() {
    return myOAuthConnector.getAccount();
  }

  @Override
  public void setCurrentAccount(@Nullable CheckiOAccount lastSavedAccount) {
    myOAuthConnector.setAccount(lastSavedAccount);
  }

  @NotNull
  protected HyperlinkAdapter createAuthorizeListener() {
    return new HyperlinkAdapter() {
      @Override
      protected void hyperlinkActivated(HyperlinkEvent event) {
        myOAuthConnector.doAuthorize(() -> {
          setLastSavedAccount(getCurrentAccount());
          updateLoginLabels();
        });
      }
    };
  }
}

File: educational-core/src/com/jetbrains/edu/learning/checkio/options/CheckiOOptions.kt
package com.jetbrains.edu.learning.checkio.options

import com.intellij.openapi.application.ApplicationManager
import com.jetbrains.edu.learning.checkio.account.CheckiOAccount
import com.jetbrains.edu.learning.checkio.connectors.CheckiOOAuthConnector
import com.jetbrains.edu.learning.checkio.connectors.CheckiOOAuthConnector.Companion.authorizationTopic
import com.jetbrains.edu.learning.settings.LoginOptions
import javax.swing.event.HyperlinkEvent

abstract class CheckiOOptions protected constructor(private val myOAuthConnector: CheckiOOAuthConnector) : LoginOptions<CheckiOAccount>() {
  override fun getCurrentAccount(): CheckiOAccount? {
    return myOAuthConnector.account
  }

  override fun setCurrentAccount(account: CheckiOAccount?) {
    myOAuthConnector.account = account
    val messageBus = ApplicationManager.getApplication().messageBus
    if (account != null) {
      messageBus.syncPublisher(authorizationTopic).userLoggedIn()
    }
    else {
      messageBus.syncPublisher(authorizationTopic).userLoggedOut()
    }
  }

  override fun createAuthorizeListener(): LoginListener {
    return object : LoginListener() {
      override fun authorize(e: HyperlinkEvent?) {
        myOAuthConnector.doAuthorize(Runnable {
          lastSavedAccount = getCurrentAccount()
          updateLoginLabels()
        })
      }
    }
  }
}